pkg/cli/upgrade/upgrade.go

```diff
@@ -0,0 +1,432 @@
+	ctx context.Context, currentEverestVersion *goversion.Version,
+) (*goversion.Version, *version.MetadataVersion, error) {
+	req, err := u.versionService.GetEverestMetadata(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	upgradeTo, meta := u.findNextMinorVersion(req, currentEverestVersion)
+	if upgradeTo == nil {
+		upgradeTo = currentEverestVersion
+	}
+
+	// Find the latest patch version for the given minor version.
+	for _, v := range req.GetVersions() {
+		ver, err := goversion.NewVersion(v.GetVersion())
+		if err != nil {
+			u.l.Debugf("Could not parse version %s. Error: %s", v.GetVersion(), err)
+			continue
+		}
+
+		if currentEverestVersion.GreaterThanOrEqual(ver) {
+			continue
+		}
+
+		// Select the latest patch version for the same major and minor version.
+		verSeg := ver.Segments()
+		uSeg := upgradeTo.Segments()
+		if len(verSeg) >= 3 && len(uSeg) >= 3 && verSeg[0] == uSeg[0] && verSeg[1] == uSeg[1] {
+			if verSeg[2] <= uSeg[2] {
+				continue
+			}
+			upgradeTo = ver
+			meta = v
+			continue
+		}
+	}
+
+	if upgradeTo == nil || meta == nil {
+		return nil, nil, ErrNoUpdateAvailable
+	}
+
+	return upgradeTo, meta, nil
+}
+
+func (u *Upgrade) findNextMinorVersion(
+	req *version.MetadataResponse, currentEverestVersion *goversion.Version,
+) (*goversion.Version, *version.MetadataVersion) {
+	var (
+		upgradeTo *goversion.Version
+		meta      *version.MetadataVersion
+	)
+
+	for _, v := range req.GetVersions() {
+		ver, err := goversion.NewVersion(v.GetVersion())
+		if err != nil {
+			u.l.Debugf("Could not parse version %s. Error: %s", v.GetVersion(), err)
+			continue
+		}
+
+		if currentEverestVersion.GreaterThanOrEqual(ver) {
+			continue
+		}
+
+		verSeg := ver.Segments()
+		evSeg := currentEverestVersion.Segments()
+		if len(verSeg) >= 3 && len(evSeg) >= 3 && verSeg[0] == evSeg[0] && verSeg[1] == evSeg[1] {
+			continue
+		}
+
+		if upgradeTo == nil {
+			upgradeTo = ver
+			meta = v
+			continue
+		}
+
+		if upgradeTo.GreaterThan(ver) {
+			upgradeTo = ver
+			meta = v
+			continue
+		}
+	}
+
+	return upgradeTo, meta
+}
+
+func (u *Upgrade) verifyRequirements(ctx context.Context, meta *version.MetadataVersion) error {
+	supVer, err := common.NewSupportedVersion(meta)
+	if err != nil {
+		return err
+	}
+
+	if err := u.checkRequirements(ctx, supVer); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (u *Upgrade) checkRequirements(ctx context.Context, supVer *common.SupportedVersion) error {
+	// TODO: olm, catalog to be implemented.
+
+	// cli version check.
+	if cliVersion.Version != "" {
+		u.l.Infof("Checking cli version requirements")
+		cli, err := goversion.NewVersion(cliVersion.Version)
+		if err != nil {
+			return errors.Join(err, fmt.Errorf("invalid cli version %s", cliVersion.Version))
+		}
+
+		// if !supVer.Cli.Check(cli.Core()) {
+		// 	return fmt.Errorf(
+		// 		"cli version %q does not meet minimum requirements of %q",
+		// 		cli, supVer.Cli.String(),
+		// 	)
+		// }
```

Comment:
Do we need to run Everest 1.3.x before migrating to 1.4.0?

Reply:
I had this commented out only for some local testing. I forgot to uncomment it. Thanks for noticing :) 

> Do we need to run Everest 1.3.x before migrating to 1.4.0?

but to answer this, no they don't need to.. they can directly install 1.4.0

Reply:
if you're talking about upgrades, I don't think we allow skipping versions. So you can upgrade to 1.4.0 only if you're on 1.3.x (or else you can directly install 1.4.0)